ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,315 @@
+//! |        7 |        7 |        3 |        3 | ... |        1 |
+//! |        8 |        6 |        4 |        3 | ... |        1 |
+//! |       10 |        5 |        4 |        3 | ... |        1 |
+//!
+//! As you can see, the bucket contents are not strictly sorted, but this is not an issue.
+//! We only need buckets with approximately equal queue sizes to ensure efficient parallel execution.
+//! The entire queue is processed in a single runtime instance in one go, so prioritizing larger queues improves efficiency.
+//!
+//! Once all program queues have been processed, we deterministically merge journals and handle them.
+//! After that, we repeat the process until no more messages remain
+//! or we run out of processing time/gas allowance (to be implemented).
+//!
+//! **High-level overview of the algorithm:**
+//!
+//!   1. Split programs into buckets based on their queue sizes.
+//!   2. Execute program queues in parallel using runtime instances per program.
+//!   3. Merge journals and handle them deterministically.
+//!   4. Repeat steps 1-3 until no messages are processed, or we run out of processing time/gas allowance.
+//!
+//! ---
+//!
+//! ## Simplest Bucket Splitting Algorithm
+//!
+//! A basic bucket-splitting algorithm is implemented as follows:
+//!
+//!   1. First, calculate a temporary bucket index based on the program queue size.
+//!   2. Next, store the required data in a temporary bucket list.
+//!   3. Repeat this process for all programs.
+//!   4. Since the number of elements in each temporary bucket is random, they need to be redistributed
+//!      to ensure all final buckets contain an equal number of elements.
+//!      To achieve this, we first merge all temporary bucket lists sequentially
+//!      and then redistribute the data according to the expected bucket size.
+//!
+//! ---
+//!
+//! ## Future Improvements
+//!
+//! Currently, the bucket partitioning algorithm is simple and does not consider a program’s execution time.
+//! In the future, we could introduce a weight multiplier to the queue size to improve partitioning efficiency.
+//! This weight multiplier could be calculated based on program execution time statistics.
+
+use crate::host::{InstanceCreator, InstanceWrapper};
 use ethexe_db::{CodesStorage, Database};
-use ethexe_runtime_common::{InBlockTransitions, JournalHandler, TransitionController};
-use gear_core::ids::ProgramId;
-use gprimitives::H256;
-use std::collections::BTreeMap;
-use tokio::sync::{mpsc, oneshot};
-
-enum Task {
-    Run {
-        program_id: ProgramId,
-        state_hash: H256,
-        result_sender: oneshot::Sender<(Vec<JournalNote>, Option<Origin>)>,
-    },
-}
+use ethexe_runtime_common::{
+    InBlockTransitions, JournalHandler, ProgramJournals, TransitionController,
+};
+use gprimitives::{ActorId, H256};
+use std::iter;
 
-pub fn run(
-    config: &ProcessorConfig,
+pub async fn run(
+    virtual_threads: usize,
```

Comment:
now this threads are not so virtual (because of spawn_blocking) and also we should remove worker_threads_override then

Reply:
I think "virtual threads" here has the same meaning as CPU virtual threads/cores, etc.

Reply:
I don't heard that somebody called cores threads as `virtual`

Reply:
I don't mind,
 do you have idea how it should be named?

Reply:
Maybe we should consider something more specific like: queues_processing_threads

Reply:
Because actually this is queues processing only parallelism

Reply:
Renamed